c2rust-analyze/src/util.rs

```diff
@@ -305,3 +306,44 @@ pub fn lty_project<'tcx, L: Debug>(
         ProjectionElem::Downcast(..) => todo!("type_of Downcast"),
     }
 }
+
+/// Determine if two types are safe to transmute to each other.
+///
+/// Safe transmutability is difficult to check abstractly,
+/// so here it is limited to integer types of the same size
+/// (but potentially different signedness).
+///
+/// Thus, [`true`] means it is definitely transmutable,
+/// while [`false`] means it may not be transmutable.
+pub fn are_transmutable<'tcx>(a: Ty<'tcx>, b: Ty<'tcx>) -> bool {
+    let transmutable_ints = {
+        use IntTy::*;
+        use UintTy::*;
+        match (a.kind(), b.kind()) {
+            (ty::Uint(u), ty::Int(i)) | (ty::Int(i), ty::Uint(u)) => {
+                matches!((u, i), |(Usize, Isize)| (U8, I8)
+                    | (U16, I16)
+                    | (U32, I32)
+                    | (U64, I64))
```

Comment:
I hate this formats as a closure.

Reply:
Yep, I hate it, too, but I couldn't figure out a way around `rustfmt`.

Reply:
It makes it look like a closure.

Reply:
Does removing the leading `|` make rustfmt do something more reasonable here?  Failing that, can you make rustfmt ignore this bit of code?  The closure-style formatting is actively misleading.

Reply:
Oh actually yes!  Without the leading `|` it's like this:

```rust
                matches!(
                    (u, i),
                    (Usize, Isize) | (U8, I8) | (U16, I16) | (U32, I32) | (U64, I64)
                )
```

Much better.

Reply:
Fixed in 65e51406868ad4829d8c93a12ba4743c16c4821b.